toppra/dracula/topp_dracula_interface.py

```diff
@@ -1,152 +1,72 @@
 import toppra.algorithm as algo
-import numpy as np
-import matplotlib.pyplot as plt
-
-logging.basicConfig()
+from toppra import constraint, interpolator
 
 from .zero_acceleration_start_end import ZeroAccelerationAtStartAndEnd
 
+ta.setup_logging("WARNING")
+
 
 def RunTopp(
-    knots_ext,
-    vlim,
-    alim,
-    robot_command_rate,
-    return_spline_parameters=False,
-    topp_breaks_count=1001,
-    debug_active=False,
+    waypts,  # ndarray, (N, dof)
+    vlim,  # ndarray, (dof, 2)
+    alim,  # ndarray, (dof 2)
+    # max_grid_err=1e-4,
+    return_cspl=False,
 ):
-    # robot_command_rate is not used if return_spline_parameters==True
-    if not return_spline_parameters:
-        assert robot_command_rate
-
-    knots = copy.deepcopy(knots_ext)
-    if debug_active:
-        print("Knots size: ", knots.shape)
-    ta.setup_logging("INFO")  # causing issues?
-    # Parameters
-    N_samples = knots_ext.shape[0]
-    dof = knots_ext.shape[1]
-
-    s_vector = np.linspace(0, 1, N_samples)
-    if debug_active:
-        print("num way points: %d" % N_samples)
-        print("num DOF: %d" % dof)
-    if N_samples < 10:
-        print("Error: Not enough knot points!")  # exact number tbd
-        return
-
-    path = ta.SplineInterpolator(s_vector, knots, bc_type="clamped")
-    if debug_active:
-        print("yay, we made a path")
+    N_samples = waypts.shape[0]
+    # initial x for toppra's path, essentially normalised time on x axis
+    x_max = 2.5 - 2.35 * np.exp(-0.015 * N_samples)  # empirical fit
+    x = np.linspace(0, x_max, N_samples)
+    # specifying natural here doensn't make a difference
+    # toppra only produces clamped cubic splines
+    path = ta.SplineInterpolator(x, waypts.copy(), bc_type="clamped")
     pc_vel = constraint.JointVelocityConstraint(vlim)
+    # Can be either Collocation (0) or Interpolation (1). Interpolation gives
+    # more accurate results with slightly higher computational cost
     pc_acc = constraint.JointAccelerationConstraint(
         alim, discretization_scheme=constraint.DiscretizationType.Interpolation
     )
-
+    # use the default gridpoints=None to let interpolator.propose_gridpoints
+    # calculate grid that sufficiently covers the path.
+    # this ensures the instance is controllable and avoids error:
+    #     "An error occurred when computing controllable velocities.
+    #     The path is not controllable, or is badly conditioned.
+    #     Error: Instance is not controllable"
+    # if using clamped as boundary condition, the default gridpoints error
+    # 1e-3 is ok and we don't need to calculate gridpoints
+    # boundary condition "natural" especially needs support by smaller error
+    # gridpoints = interpolator.propose_gridpoints(
+    #     path, max_err_threshold=max_grid_err
+    # )
     instance = algo.TOPPRA(
         [pc_vel, pc_acc],
         path,
-        gridpoints=np.linspace(0, 1, int(topp_breaks_count)),
+        # gridpoints=gridpoints,
```

Comment:
@duncancmt-dexai 
AFAIK, this is the important change in this PR, or at least the change that fixes our current problem

Reply:
and L20

Reply:
gotcha